api/python/tests/integration/test_packages.py

```diff
@@ -1410,3 +1424,98 @@ def test_resolve_hash(self):
 
         with pytest.raises(QuiltException, match='Found multiple matches'):
             Package.resolve_hash(pkg_name, LOCAL_REGISTRY, hash_prefix)
+
+
+# The following tests were moved out of the PackageTest class to enable parametrization.
+# see (https://docs.pytest.org/en/latest/unittest.html#pytest-features-in-unittest-testcase-subclasses)
+@pytest.mark.parametrize(
+    'target_dir, update_policy, expected_one_byte, expected_two_byte, expected_three_byte',
+    [
+        ('/', None, b'one', b'2', b'3'),
+        ('/', 'incoming', b'one', b'2', b'3'),
+        ('/', 'existing', b'1', b'2', b'3'),
+        ('sub/', None, b'1', b'two', b'three'),
+        ('sub/', 'incoming', b'1', b'two', b'three'),
+        ('sub/', 'existing', b'1', b'2', b'3'),
+        pytest.param('/', 'bad_policy', b'1', b'2', b'3', marks=pytest.mark.xfail(raises=ValueError)),
+    ]
+)
+def test_set_dir_update_policy(
+    target_dir: str,
+    update_policy: str,
+    expected_one_byte: bytes,
+    expected_two_byte: bytes,
+    expected_three_byte: bytes
+):
+    """Verify building a package with update policy. """
+    data_dir = pathlib.Path(__file__).parent / "data"
+    nested_dir = data_dir / 'nested'
+    pkg = Package()
+    pkg.set_dir("/", nested_dir, meta={'name': 'test_meta'})
+    assert 'one.txt' in pkg.keys()
+    assert 'two.txt' not in pkg.keys()
+    assert 'three.txt' not in pkg.keys()
+    assert pkg.meta == {'name': 'test_meta'}
+
+    nested_dir_2 = data_dir / 'nested2'
+    if update_policy:
+        pkg.set_dir(target_dir, nested_dir_2, update_policy=update_policy)
+    else:
+        pkg.set_dir(target_dir, nested_dir_2)
+    assert pkg['one.txt'].get_bytes() == expected_one_byte
+    assert pkg['sub/two.txt'].get_bytes() == expected_two_byte
+    assert pkg['sub/three.txt'].get_bytes() == expected_three_byte
+    if target_dir == '/':
```

Comment:
If instead you do `target_dir=''` in parameterize then everything just works, including `p.set_dir("", whatever)`. And a few points of coverage you are missing:
* `sub/` is an existing directory in `nested`; you should really show that it exists before doing the `set_dir`, because what we want to prove is that the package does the right thing based on what's on disk; at the very least you can make a comment about what you expect to be in `nested/sub` and then prove that it gets captured
* The tests prove that certain keys have expected values, but how do we know that we haven't added new keys? We don't. Add something like `assert set(pkg.keys()) == set(["known", "good", "inputs"])`.

Here's how to think about this whole unit test. You are merging A, B:
* Prove A and B are in a given initial state
* Do the merge
* Prove that only the expected adds/modifies/deletes happened